app/src/main/java/io/appium/uiautomator2/handler/W3CActions.java

```diff
@@ -0,0 +1,276 @@
+            for (final InputEventParams eventParam : eventParams) {
+                if (eventParam instanceof KeyInputEventParams) {
+                    final int keyCode = ((KeyInputEventParams) eventParam).keyCode;
+                    final int keyAction = ((KeyInputEventParams) eventParam).keyAction;
+                    if (keyCode > KeyEvent.getMaxKeyCode()) {
+                        if (keyAction == KeyEvent.ACTION_DOWN) {
+                            depressedMetaKeys.add(keyCode);
+                        } else {
+                            depressedMetaKeys.remove(keyCode);
+                        }
+                    } else {
+                        result &= injectEventSync(new KeyEvent(startTimestamp + eventParam.startDelta,
+                                SystemClock.uptimeMillis(), keyAction, keyCode, metaKeysToState(depressedMetaKeys),
+                                KeyCharacterMap.VIRTUAL_KEYBOARD, 0, 0, InputDevice.SOURCE_KEYBOARD));
+                    }
+                } else if (eventParam instanceof MotionInputEventParams) {
+                    final int inputSource = toolTypeToInputSource(((MotionInputEventParams) eventParam).properties.toolType);
+                    final List<MotionInputEventParams> events = (motionParamsByInputSource.get(inputSource) == null) ?
+                            new ArrayList<MotionInputEventParams>() :
+                            motionParamsByInputSource.get(inputSource);
+                    events.add((MotionInputEventParams) eventParam);
+                    motionParamsByInputSource.put(inputSource, events);
+                }
+            }
+
+            for (int i = 0; i < motionParamsByInputSource.size(); i++) {
+                final int inputSource = (int) motionParamsByInputSource.keyAt(i);
+                final List<MotionInputEventParams> motionEventsParams = motionParamsByInputSource.valueAt(i);
+                final PointerProperties[] nonHoveringProps = filterPointerProperties(motionEventsParams,
+                        false);
+                final PointerProperties[] hoveringProps = filterPointerProperties(motionEventsParams,
+                        true);
+                final PointerCoords[] nonHoveringCoords = filterPointerCoordinates(motionEventsParams,
+                        false);
+                final PointerCoords[] hoveringCoords = filterPointerCoordinates(motionEventsParams,
+                        true);
+
+                for (final MotionInputEventParams motionEventParams : motionEventsParams) {
+                    final int actionCode = motionEventParams.actionCode;
+                    Integer upDownBalance = motionEventsBalanceByInputSource.get(inputSource);
+                    if (upDownBalance == null) {
+                        upDownBalance = 0;
+                    }
+                    switch (actionCode) {
+                        case MotionEvent.ACTION_DOWN: {
+                            ++upDownBalance;
```

Comment:
To my preference, this is a lot of code in a case statement. I'd rather these be broken out into methods.

Reply:
I thought about this, but there are too many variables shared between blocks. Do you have any concrete proposals?

Reply:
The method is not long, no. But the switch statement is difficult to read. Things can be simplified, even if not broken out. This case, for instance, could be (as far as I can tell)
```java
upDownBalance++;
motionEventsBalanceByInputSource.put(inputSource, upDownBalance);
int pointerAction = upDownBalance == 1 ? MotionEvent.ACTION_DOWN
                                       : getPointerAction(MotionEvent.ACTION_POINTER_DOWN, upDownBalance - 1);
result &= injectEventSync(MotionEvent.obtain(startTimestamp + motionEventParams.startDelta,
        SystemClock.uptimeMillis(), pointerAction,
        upDownBalance, nonHoveringProps, nonHoveringCoords, metaKeysToState(depressedMetaKeys),
        motionEventParams.button, 1, 1, 0, 0, inputSource, 0));
```

Reply:
Thanks, fixed